pkg/servicenameresolutiondetector/coredns/detector.go

```diff
@@ -0,0 +1,290 @@
+		cacheSynced:         []cache.InformerSynced{nodeInformer.Informer().HasSynced},
+		eventBroadcaster:    broadcaster,
+		eventRecorder:       recorder,
+		queue:               workqueue.NewRateLimitingQueueWithConfig(workqueue.DefaultControllerRateLimiter(), workqueue.RateLimitingQueueConfig{Name: name}),
+		lec: leaderelection.LeaderElectionConfig{
+			Lock:          rl,
+			LeaseDuration: baselec.LeaseDuration.Duration,
+			RenewDeadline: baselec.RenewDeadline.Duration,
+			RetryPeriod:   baselec.RetryPeriod.Duration,
+		}}, nil
+}
+
+// Run starts the detector.
+func (d *Detector) Run(ctx context.Context) {
+	defer runtime.HandleCrash()
+
+	d.eventBroadcaster.StartStructuredLogging(0)
+	d.eventBroadcaster.StartRecordingToSink(&v1.EventSinkImpl{Interface: d.memberClusterClient.CoreV1().Events("")})
+	defer d.eventBroadcaster.Shutdown()
+
+	defer d.queue.ShutDown()
+
+	logger := klog.FromContext(ctx)
+	logger.Info("Starting coredns detector")
+	defer logger.Info("Shutting down coredns detector")
+
+	if !cache.WaitForCacheSync(ctx.Done(), d.cacheSynced...) {
+		return
+	}
+
+	go func() {
+		wait.Until(func() {
+			defer runtime.HandleCrash()
+
+			observed := lookupOnce(logger)
+			curr, err := d.conditionStore.Load(d.nodeName)
+			if err != nil {
+				d.eventRecorder.Eventf(localReference, corev1.EventTypeWarning, "LoadCorednsConditionFailed", "failed to load condition: %v", err)
+				return
+			}
+
+			cond := d.conditionCache.ThresholdAdjustedCondition(d.nodeName, curr, observed)
+			if err = d.conditionStore.Store(d.nodeName, cond); err != nil {
+				d.eventRecorder.Eventf(localReference, corev1.EventTypeWarning, "StoreCorednsConditionFailed", "failed to store condition: %v", err)
+				return
+			}
+			d.queue.Add(0)
+		}, d.periodSeconds, ctx.Done())
+	}()
+
+	if d.lec.Lock != nil {
+		d.lec.Callbacks = leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				wait.UntilWithContext(ctx, d.worker, time.Second)
+			},
+			OnStoppedLeading: func() {
+				logger.Error(nil, "leader election lost")
+				klog.FlushAndExit(klog.ExitFlushTimeout, 1)
+			},
+		}
+		leaderelection.RunOrDie(ctx, d.lec)
+	} else {
+		wait.UntilWithContext(ctx, d.worker, time.Second)
+	}
+}
+
+func (d *Detector) worker(ctx context.Context) {
+	for d.processNextWorkItem(ctx) {
+	}
+}
+
+func (d *Detector) processNextWorkItem(ctx context.Context) bool {
+	key, quit := d.queue.Get()
+	if quit {
+		return false
+	}
+	defer d.queue.Done(key)
+
+	if err := d.sync(ctx); err != nil {
+		runtime.HandleError(fmt.Errorf("failed to sync corendns condition to control plane, requeuing: %v", err))
+		d.queue.AddRateLimited(key)
+	} else {
+		d.queue.Forget(key)
+	}
+	return true
+}
+
+func lookupOnce(logger klog.Logger) *metav1.Condition {
+	logger.Info("lookup service name once")
+	observed := &metav1.Condition{Type: condType, LastTransitionTime: metav1.Now()}
+	if _, err := net.LookupHost("kubernetes.default"); err != nil {
```

Comment:
Fault detection is a relatively complex matter, and it may not be as simple as this test.
Moreover, the cost of maintaining such a new component can be high.

I think what you have here is more like an example, used to demonstrate how `Remedy` works.

 Karmada should focus on the generality and scalability of `Remedy`, not how faults are detected.

Reply:
Agree.